src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,394 @@
+
+    new(key, value) =
+        {
+            Key = key
+            Value = value
+            AccessCount = 0L
+            Node = Unchecked.defaultof<_>
+        }
+
+    member this.WithNode() =
+        this.Node <- LinkedListNode(this)
+        this
+
+    member this.ReUse(key, value) =
+        this.Key <- key
+        this.Value <- value
+        this.AccessCount <- 0L
+        this
+
+    override this.ToString() = $"{this.Key}"
+
+type EntityPool<'Key, 'Value>(maximumCapacity, overCapacity: Event<_>) =
+    let pool = ConcurrentBag<CachedEntity<'Key, 'Value>>()
+    let mutable created = 0
+
+    member _.Acquire(key, value) =
+        match pool.TryTake() with
+        | true, entity -> entity.ReUse(key, value)
+        | _ ->
+            if Interlocked.Increment &created > maximumCapacity then
+                overCapacity.Trigger()
+
+            CachedEntity(key, value).WithNode()
+
+    member _.Reclaim(entity: CachedEntity<'Key, 'Value>) =
+        if pool.Count < maximumCapacity then
+            pool.Add(entity)
+
+type IEvictionQueue<'Key, 'Value> =
+    abstract member Add: CachedEntity<'Key, 'Value> -> unit
+    abstract member Update: CachedEntity<'Key, 'Value> -> unit
+    abstract member GetKeysToEvict: int -> 'Key[]
+    abstract member Remove: CachedEntity<'Key, 'Value> -> unit
+
+[<Sealed; NoComparison; NoEquality>]
+type EvictionQueue<'Key, 'Value>() =
+
+    let list = LinkedList<CachedEntity<'Key, 'Value>>()
+
+    interface IEvictionQueue<'Key, 'Value> with
+
+        member _.Add(entity: CachedEntity<'Key, 'Value>) =
+            lock list
+            <| fun () ->
+                if isNull entity.Node.List then
+                    list.AddLast(entity.Node)
+                else
+                    assert false
+
+        member _.Update(entity: CachedEntity<'Key, 'Value>) =
+            lock list
+            <| fun () ->
+                Interlocked.Increment(&entity.AccessCount) |> ignore
+
+                let node = entity.Node
+
+                // Sync between store and the eviction queue is not atomic. It might be already evicted or not yet added.
+                if node.List = list then
+                    // Just move this node to the end of the list.
+                    list.Remove(node)
+                    list.AddLast(node)
+
+        member _.GetKeysToEvict(count) =
+            lock list
+            <| fun () -> list |> Seq.map _.Key |> Seq.truncate count |> Seq.toArray
+
+        member this.Remove(entity: CachedEntity<_, _>) =
+            lock list <| fun () -> list.Remove(entity.Node)
+
+    member _.Count = list.Count
+
+    static member NoEviction =
+        { new IEvictionQueue<'Key, 'Value> with
+            member _.Add(_) = ()
+
+            member _.Update(entity) =
+                Interlocked.Increment(&entity.AccessCount) |> ignore
+
+            member _.GetKeysToEvict(_) = [||]
+            member _.Remove(_) = ()
+        }
+
+type ICacheEvents =
```

Comment:
Are events needed or should we just use `Metrics` counters directly?

Reply:
It was added for observability.
If we can use metrics for it, it serves better.

(also less worries about loose GC handles IMO)